contracts/AlephZero_SpokePool.sol

```diff
@@ -0,0 +1,135 @@
+     *    ARBITRUM-SPECIFIC CROSS-CHAIN ADMIN FUNCTIONS     *
+     ********************************************************/
+
+    /**
+     * @notice Change L2 gateway router. Callable only by admin.
+     * @param newL2GatewayRouter New L2 gateway router.
+     */
+    function setL2GatewayRouter(address newL2GatewayRouter) public onlyAdmin nonReentrant {
+        _setL2GatewayRouter(newL2GatewayRouter);
+    }
+
+    /**
+     * @notice Add L2 -> L1 token mapping. Callable only by admin.
+     * @param l2Token Arbitrum token.
+     * @param l1Token Ethereum version of l2Token.
+     */
+    function whitelistToken(address l2Token, address l1Token) public onlyAdmin nonReentrant {
+        _whitelistToken(l2Token, l1Token);
+    }
+
+    /**************************************
+     *        INTERNAL FUNCTIONS          *
+     **************************************/
+
+    function _bridgeTokensToHubPool(uint256 amountToReturn, address l2TokenAddress) internal override {
+        // If the l2TokenAddress is UDSC, we need to use the CCTP bridge.
+        if (_isCCTPEnabled() && l2TokenAddress == address(usdcToken)) {
+            _transferUsdc(hubPool, amountToReturn);
+        } else {
+            // Check that the Ethereum counterpart of the L2 token is stored on this contract.
+            address ethereumTokenToBridge = whitelistedTokens[l2TokenAddress];
+            require(ethereumTokenToBridge != address(0), "Uninitialized mainnet token");
+            //slither-disable-next-line unused-return
+            StandardBridgeLike(l2GatewayRouter).outboundTransfer(
+                ethereumTokenToBridge, // _l1Token. Address of the L1 token to bridge over.
+                hubPool, // _to. Withdraw, over the bridge, to the l1 hub pool contract.
+                amountToReturn, // _amount.
+                "" // _data. We don't need to send any data for the bridging action.
+            );
+        }
+    }
+
+    function _setL2GatewayRouter(address _l2GatewayRouter) internal {
+        l2GatewayRouter = _l2GatewayRouter;
+        emit SetL2GatewayRouter(l2GatewayRouter);
+    }
+
+    function _whitelistToken(address _l2Token, address _l1Token) internal {
+        whitelistedTokens[_l2Token] = _l1Token;
+        emit WhitelistedTokens(_l2Token, _l1Token);
+    }
+
+    // L1 addresses are transformed during l1->l2 calls.
+    // See https://developer.offchainlabs.com/docs/l1_l2_messages#address-aliasing for more information.
+    // This cannot be pulled directly from Arbitrum contracts because their contracts are not 0.8.X compatible and
+    // this operation takes advantage of overflows, whose behavior changed in 0.8.0.
+    function _applyL1ToL2Alias(address l1Address) internal pure returns (address l2Address) {
+        // Allows overflows as explained above.
+        unchecked {
+            l2Address = address(uint160(l1Address) + uint160(0x1111000000000000000000000000000000001111));
+        }
+    }
```

Comment:
@bmzig For future, we should refactor this to use the library you created. We could have that standing as a PR to include for audit at some future point in time.